apps/prairielearn/src/lib/grading.js

```diff
@@ -43,7 +43,8 @@ export function saveSubmission(submission, variant, question, variant_course, ca
   debug('saveSubmission()');
   submission.raw_submitted_answer = submission.submitted_answer;
   submission.gradable = true;
-  let questionModule, question_course, courseIssues, data, submission_id, workspace_id, zipPath;
+  const questionModule = questionServers.getModule(question.type);
```

Comment:
This can throw an error; calling a function that can throw in a callback-based function is unsafe as the error may not be correctly propagated back up.

Reply:
This can throw an error if the question type does not match the expected values, which is not expected to happen given that the type comes from a DB enum. Given that throwing this error would be a problem in the system (and not a user error), how do you want to handle this?

* Ignore the error in the question server module, send it to sentry/logger, and return a default value;
* Receive an optional callback function and call it if set in case of error;
* Have a big try/catch block here that calls the callback on error.

Reply:
IMO we don't need to think about system vs. user error in this case. The problem is that if an error is thrown but not caught (as is the case here), it actually crashes the entire server, which is unacceptable for *any* kind of error. You can reproduce this by configuring `getModule` to unconditionally throw an error.

My recommendation would be to either a) rewrite the enclosing functions as pure async/await (big lift but has to happen eventually), or b) keep the call to `quesitonServers.getModule` inside of the `async.series` flows, but use an async enclosing function so that errors are handled correctly. I'd lean towards (a) here, just to minimize changes. To be specific:

```
let questionModule;
async.series([
  async () => {
    questionModule = questionServers.getModule(...)
  },
  (callback) => {
  // ...
  }
]);

Reply:
Replaced with try-catch. Rewriting it as pure async/await would be a lot of work in this context, and I don't want to mix up the changes.

Reply:
Sorry, I left the wrong letter in my "I'd lean towards"... I agree that we should not do a wholesale rewrite in async/await.

Reply:
Sure, I can do that where `async.series` was used, but a couple of the functions did not use `async.series`. Are you ok with try/catch there?

Reply:
`try`/`catch` is fine, though you may consider changing those to use `async.series` instead too, as that often makes the future conversion to purely `async`/`await` code more straightforward. I'll leave it up to you.

Reply:
Done in the latest commits.